File: python/lira/ir_std.py
from .ir import *
from .arch import *

from dataclasses import dataclass
from enum import Enum
from typing import ClassVar


class StmtKind(str, Enum):
    INPUT = "input"
    OUTPUT = "output"
    READ = "read"
    WRITE = "write"
    OP = "op"
    ENV = "env"
    COND_ENV = "cond_env"
    CONST = "const"
    DYN_CONST = "dyn_const"


@dataclass
class StmtInput:
    kind: ClassVar[StmtKind] = StmtKind.INPUT
    id_: int

@dataclass
class StmtOutput:
    kind: ClassVar[StmtKind] = StmtKind.OUTPUT
    id_: int
    value: str

@dataclass
class StmtRead:
    kind: ClassVar[StmtKind] = StmtKind.READ
    rf: RegisterFile
    rsi: str

@dataclass
class StmtWrite:
    kind: ClassVar[StmtKind] = StmtKind.WRITE
    rf: RegisterFile
    rsi: str
    value: str

@dataclass
class StmtOp:
    kind: ClassVar[StmtKind] = StmtKind.OP
    op: Operation
    args: list[str]

@dataclass
class StmtEnv:
    kind: ClassVar[StmtKind] = StmtKind.ENV
    env: EnvironmentFunction
    args: list[str]

class CondEnv:
    kind: ClassVar[StmtKind] = StmtKind.COND_ENV
    env: EnvironmentFunction
    cond: str
    on_false: list[str]
    inputs: list[str]

@dataclass
class StmtIndex:
    pass

@dataclass
class StmtConst:
    kind: ClassVar[StmtKind] = StmtKind.CONST
    value: int

@dataclass
class StmtDynConst:
    kind: ClassVar[StmtKind] = StmtKind.DYN_CONST
    name: str

@dataclass
class StmtGather:
    value: str
    index: str
    default: str

@dataclass
class StmtFold:
    op: Operation
    args: list[str]

@dataclass
class StmtScan:
    op: Operation
    args: list[str]

@dataclass
class StmtAlias:
    semantic: Snippet
    args: list[str]

File: ruby/lira/ir_std.rb
module LIRA
  class StmtInput
    attr_reader :id_
    def initialize(id_); @id_ = id_; end
  end

  class StmtOutput
    attr_reader :id_, :value
    def initialize(id_, value); @id_ = id_; @value = value; end
  end

  class StmtRead
    attr_reader :rf, :rsi
    def initialize(rf, rsi); @rf = rf; @rsi = rsi; end
  end

  class StmtWrite
    attr_reader :rf, :rsi, :value
    def initialize(rf, rsi, value); @rf = rf; @rsi = rsi; @value = value; end
  end

  class StmtOp
    attr_reader :op, :args
    def initialize(op, args); @op = op; @args = args; end
  end

  class StmtEnv
    attr_reader :env, :args
    def initialize(env, args); @env = env; @args = args; end
  end

  class CondEnv
    attr_reader :env, :cond, :on_false, :inputs
    def initialize(env, cond, on_false, inputs); @env = env; @cond = cond; @on_false = on_false; @inputs = inputs; end
  end

  class StmtIndex; end

  class StmtConst
    attr_reader :value
    def initialize(value); @value = value; end
  end

  class StmtDynConst
    attr_reader :name
    def initialize(name); @name = name; end
  end

  class StmtGather
    attr_reader :value, :index, :default
    def initialize(value, index, default); @value = value; @index = index; @default = default; end
  end

  class StmtFold
    attr_reader :op, :args
    def initialize(op, args); @op = op; @args = args; end
  end

  class StmtScan
    attr_reader :op, :args
    def initialize(op, args); @op = op; @args = args; end
  end

  class StmtAlias
    attr_reader :semantic, :args
    def initialize(semantic, args); @semantic = semantic; @args = args; end
  end
end
